### PR TITLE
Use structs for AAD encoding

### LINKED ISSUE #371: Clarify encoding of AAD
Currently, the AAD for input shares is defined as `task_id || metadata || public_share`. The variable `public_share` is narratively introduced in 4.4.1.3 as "the public share sent to each aggregator", referring to a part of the report share. Two implementations interpreted the construction of the AAD differently, as either [including](https://github.com/cloudflare/daphne/blob/61ab97f0c07f9f4bbdb13ec666cebd22febf12ec/daphne/src/vdaf/mod.rs#L184) or [not including](https://github.com/divviup/janus/blob/42873cddb666dce39d27a2342adc3d7232b14eda/core/src/hpke.rs#L37-L49) a 32-bit length suffix preceding the public share.

While it would likely be okay to have a single variable-length byte array in the AAD without a length prefix, as the resulting AAD would still be unambiguous, I think it would be prudent to change this to explicitly include a length prefix, to simplify future changes.

We should define new structs for both kinds of AAD strings, and specify that the AAD is the result of encoding the corresponding struct. For example:

```
struct {
  TaskID task_id;
  ReportMetadata metadata;
  opaque public_share<0..2^32-1>;
} ReportShareAAD;
```

### PR DESCRIPTION
**Pull request by divergentdave**
This revises the sections dealing with encryption and decryption of input shares and aggregate shares, rewriting the construction of each AAD input to be in terms of RFC 8446-style structs instead of byte string concatenation. This fixes #371.

### CODE REVIEW

**Review by cjpatton — APPROVED** [review]
Solid implementation, and a clear win 👍

**Review by tgeoghegan — APPROVED** [review]

File: draft-ietf-ppm-dap.md [hunk 1]
```diff
@@ -1636,26 +1646,36 @@ done as follows:
...
 enc, payload = SealBase(pk, "dap-02 aggregate share" || server_role || 0x00,
-  AggregateShareReq.task_id || AggregateShareReq.batch_selector, agg_share)
+  agg_share_aad, agg_share)
 ~~~
 
 where `pk` is the HPKE public key encoded by the collector's HPKE key,
-`server_role` is `0x02` for the leader and `0x03` for a helper.
+`server_role` is `0x02` for the leader and `0x03` for a helper, and
+`agg_share_aad` is constructed from the `task_id` and `batch_selector` fields
+of the `AggregateShareReq`, with the following format:
```
Comment (ekr) [inline draft-ietf-ppm-dap.md:1655]:
```suggestion
`server_role` is `0x02` for the leader and `0x03` for a helper, and
`agg_share_aad` is a value of type `AggregateShareAAD` with its values set from the corresponding fields of the `AggregateShareReq`.
```

File: draft-ietf-ppm-dap.md [hunk 2]
```diff
@@ -1636,26 +1646,36 @@ done as follows:
...
-request, and `server_role` is the role of the server that sent the aggregate
-share (`0x02` for the leader and `0x03` for the helper). The value of
-`batch_selector` is computed by the Collector from its query and the response to
-its query:
+where `sk` is the HPKE secret key, `server_role` is the role of the server that
+sent the aggregate share (`0x02` for the leader and `0x03` for the helper), and
+`agg_share_aad` is an `AggregateShareAAD` message constructed from the task ID
+in the collect request and a batch selector. The value of the batch selector
+used in `agg_share_aad` is computed by the Collector from its query and the
+response to its query:
```
Comment (ekr) [inline draft-ietf-ppm-dap.md:1678]:
```suggestion
response to its query as follows:
```

**Review by ekr — CHANGES_REQUESTED** [review]
Generally LGTM.

File: draft-ietf-ppm-dap.md [hunk 3]
```diff
@@ -810,13 +810,22 @@ as specified by the VDAF. It then encrypts each input share as follows:
...
+`input_share` is the Aggregator's input share, and `input_share_aad` is an
+encoded message of type `InputShareAAD`, constructed from the same values as
+the corresponding fields in the report:
+
+~~~
+struct {
+  TaskID task_id;
+  ReportMetadata metadata;
+  opaque public_share<0..2^32-1>;
+} InputShareAAD;
```
Comment (wangshan) [inline draft-ietf-ppm-dap.md:827]: nit: if we follow he same camel case convention of types, this should be `InputShareAad`. Same applies to `AggregateShareAad`

**Review by wangshan — APPROVED** [review]

**Comment by divergentdave** [comment]
I squashed this back into two commits and then rebased to fix some conflicts.

**Comment by cjpatton** [comment]
@ekr would you mind taking a quick look?

**Review by ekr — APPROVED** [review]
LGTM

**Pull request merged**